autotest/utilities/test_ogr2ogr_lib.py

```diff
@@ -2187,3 +2188,96 @@ def my_handler(errorClass, errno, msg):
     assert (
         out_lyr.GetFeatureCount() == (limit if limit else src_lyr.GetFeatureCount()) + 2
     )
+
+
+###############################################################################
+# Test JSON types roundtrip
+
+
+@pytest.mark.require_driver("GeoJSON")
+@pytest.mark.require_driver("GPKG")
+def test_json_types(tmp_vsimem):
+    """Test JSON types"""
+
+    def test_extended_types(lyr):
+        assert lyr.GetFeatureCount() == 1
+        f = lyr.GetNextFeature()
+
+        fd = f.GetFieldDefnRef(0)
+        assert fd.GetType() == ogr.OFTString
+        assert fd.GetSubType() == ogr.OFSTNone
+
+        fd = f.GetFieldDefnRef(1)
+        assert fd.GetType() == ogr.OFTIntegerList
+        assert fd.GetSubType() == ogr.OFSTNone
+
+        fd = f.GetFieldDefnRef(2)
+        assert fd.GetType() == ogr.OFTString
+        assert fd.GetSubType() == ogr.OFSTJSON
+
+        fd = f.GetFieldDefnRef(3)
+        assert fd.GetType() == ogr.OFTInteger
+        assert fd.GetSubType() == ogr.OFSTNone
+
+    def test_types(lyr):
+        assert lyr.GetFeatureCount() == 1
+        f = lyr.GetNextFeature()
+
+        fd = f.GetFieldDefnRef(0)
+        assert fd.GetType() == ogr.OFTString
+        assert fd.GetSubType() == ogr.OFSTNone
+
+        fd = f.GetFieldDefnRef(1)
+        assert fd.GetType() == ogr.OFTString
+        assert fd.GetSubType() == ogr.OFSTJSON
+
+        fd = f.GetFieldDefnRef(2)
+        assert fd.GetType() == ogr.OFTString
+        assert fd.GetSubType() == ogr.OFSTJSON
+
+        fd = f.GetFieldDefnRef(3)
+        assert fd.GetType() == ogr.OFTInteger
+        assert fd.GetSubType() == ogr.OFSTNone
+
+    with gdal.ExceptionMgr(useExceptions=True):
+
+        with tempfile.TemporaryDirectory() as temp_dir:
+            src = os.path.join(temp_dir, "test_json.geojson")
+            dst = os.path.join(temp_dir, "test_json.gpkg")
+
+            with open(src, "w+") as f:
+                f.write(
```

Comment:
should rather be something like (with indentation fixed):

```suggestion
            src = str(tmp_vsimem / "test_json.geojson")
            dst = str(tmp_vsimem / "test_json.gpkg")

           gdal.FileFromMemBuffer(src,
```

Reply:
done